File: app/src/main/java/com/stravawidget/data/StravaRepository.kt
package com.stravawidget.data

import android.content.Context
import com.stravawidget.BuildConfig
import com.stravawidget.api.Activity
import com.stravawidget.api.StravaClient
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit

class StravaRepository(context: Context) {

    private val tokenStorage = TokenStorage(context)
    private val activityCache = ActivityCache(context)
    private val api = StravaClient.api

    val isLoggedIn: Boolean
        get() = tokenStorage.isLoggedIn

    suspend fun exchangeCodeForTokens(code: String): Result<Unit> {
        return try {
            val response = api.exchangeToken(
                clientId = BuildConfig.STRAVA_CLIENT_ID,
                clientSecret = BuildConfig.STRAVA_CLIENT_SECRET,
                code = code
            )
            tokenStorage.saveTokens(
                accessToken = response.access_token,
                refreshToken = response.refresh_token,
                expiresAt = response.expires_at
            )
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private suspend fun refreshTokenIfNeeded(): Boolean {
        if (!tokenStorage.isTokenExpired) return true

        val refreshToken = tokenStorage.refreshToken ?: return false

        return try {
            val response = api.refreshToken(
                clientId = BuildConfig.STRAVA_CLIENT_ID,
                clientSecret = BuildConfig.STRAVA_CLIENT_SECRET,
                refreshToken = refreshToken
            )
            tokenStorage.saveTokens(
                accessToken = response.access_token,
                refreshToken = response.refresh_token,
                expiresAt = response.expires_at
            )
            true
        } catch (e: Exception) {
            false
        }
    }

    suspend fun syncActivities(): Result<Map<String, Int>> {
        if (!isLoggedIn) return Result.failure(Exception("Not logged in"))

        if (!refreshTokenIfNeeded()) {
            return Result.failure(Exception("Failed to refresh token"))
        }

        val accessToken = tokenStorage.accessToken ?: return Result.failure(Exception("No access token"))

        return try {
            val after = getStartOfPeriod()
            val activities = api.getActivities(
                authorization = "Bearer $accessToken",
                after = after
            )

            val levels = calculateLevels(activities)
            activityCache.saveActivityLevels(levels)

            Result.success(levels)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    fun getCachedActivityLevels(): Map<String, Int> {
        return activityCache.getActivityLevels()
    }

    private fun getStartOfPeriod(): Long {
        val today = LocalDate.now()
        val startDate = today.minusDays(91)
        return startDate.atStartOfDay(ZoneId.systemDefault()).toEpochSecond()
    }

    private fun calculateLevels(activities: List<Activity>): Map<String, Int> {
        // Group activities by date and sum moving_time
        val minutesByDate = mutableMapOf<String, Int>()

        for (activity in activities) {
            // Extract date from start_date_local (first 10 chars: YYYY-MM-DD)
            val date = activity.start_date_local.take(10)
            val minutes = activity.moving_time / 60
            minutesByDate[date] = (minutesByDate[date] ?: 0) + minutes
        }

        // Convert minutes to levels
        return minutesByDate.mapValues { (_, minutes) ->
            when {
                minutes == 0 -> 0
                minutes < 30 -> 1
                minutes < 60 -> 2
                minutes < 90 -> 3
                else -> 4
            }
        }
    }

    fun logout() {
        tokenStorage.clear()
        activityCache.clear()
    }

    companion object {
        private const val WEEKS_TO_SHOW = 13
        private const val DAYS_TO_SHOW = WEEKS_TO_SHOW * 7

        fun getGridDates(): List<LocalDate?> {
            val today = LocalDate.now()
            val currentDayOfWeek = today.dayOfWeek.value % 7 // Sunday = 0

            // End date is Saturday of current week
            val endDate = today.plusDays((6 - currentDayOfWeek).toLong())
            // Start date is Sunday 12 weeks before
            val startDate = endDate.minusDays(DAYS_TO_SHOW.toLong() - 1)

            val dates = mutableListOf<LocalDate?>()

            // Build grid: 7 rows (days) x 13 columns (weeks)
            // Row 0 = Sunday, Row 6 = Saturday
            for (dayOfWeek in 0..6) {
                for (week in 0 until WEEKS_TO_SHOW) {
                    val date = startDate.plusDays((week * 7 + dayOfWeek).toLong())
                    // Hide future dates
                    if (date.isAfter(today)) {
                        dates.add(null)
                    } else {
                        dates.add(date)
                    }
                }
            }

            return dates
        }

        fun formatDate(date: LocalDate): String {
            return date.format(DateTimeFormatter.ISO_LOCAL_DATE)
        }
    }
}

File: ios/StravaWidget/Data/StravaRepository.swift
import Foundation
import WidgetKit

final class StravaRepository {
    static let shared = StravaRepository()

    private let tokenStorage = TokenStorage.shared
    private let activityCache = ActivityCache.shared
    private let client = StravaClient.shared

    private init() {}

    var isLoggedIn: Bool {
        tokenStorage.isLoggedIn
    }

    // MARK: - Token Management

    func saveTokens(from response: TokenResponse) {
        tokenStorage.saveTokens(from: response)
    }

    private func refreshTokenIfNeeded() async -> Bool {
        guard tokenStorage.isTokenExpired else { return true }
        guard let refreshToken = tokenStorage.refreshToken else { return false }

        do {
            let response = try await client.refreshToken(refreshToken)
            saveTokens(from: response)
            return true
        } catch {
            return false
        }
    }

    // MARK: - Sync Activities

    @discardableResult
    func syncActivities() async throws -> [String: Int] {
        guard isLoggedIn else {
            throw RepositoryError.notLoggedIn
        }

        guard await refreshTokenIfNeeded() else {
            throw RepositoryError.tokenRefreshFailed
        }

        guard let accessToken = tokenStorage.accessToken else {
            throw RepositoryError.noAccessToken
        }

        let after = getStartOfPeriod()
        let activities = try await client.getActivities(accessToken: accessToken, after: after)
        let levels = calculateLevels(activities)

        activityCache.saveActivityLevels(levels)

        // Reload widget timelines
        WidgetCenter.shared.reloadTimelines(ofKind: "StravaActivityWidget")

        return levels
    }

    func getCachedActivityLevels() -> [String: Int] {
        activityCache.getActivityLevels()
    }

    // MARK: - Level Calculation

    private func getStartOfPeriod() -> TimeInterval {
        let calendar = Calendar.current
        let today = Date()
        let startDate = calendar.date(byAdding: .day, value: -Constants.totalDays, to: today)!
        return calendar.startOfDay(for: startDate).timeIntervalSince1970
    }

    private func calculateLevels(_ activities: [StravaActivity]) -> [String: Int] {
        var minutesByDate: [String: Int] = [:]

        for activity in activities {
            let date = String(activity.startDateLocal.prefix(10))
            let minutes = activity.movingTime / 60
            minutesByDate[date, default: 0] += minutes
        }

        return minutesByDate.mapValues { minutes in
            ActivityLevel.from(minutes: minutes).rawValue
        }
    }

    // MARK: - Grid Date Calculation

    static func getGridDates() -> [Date?] {
        let calendar = Calendar.current
        let today = Date()

        // Get current day of week (Sunday = 1 in Calendar)
        let currentDayOfWeek = calendar.component(.weekday, from: today) - 1

        // End date is Saturday of current week
        let endDate = calendar.date(byAdding: .day, value: 6 - currentDayOfWeek, to: today)!

        // Start date is Sunday 12 weeks before
        let startDate = calendar.date(byAdding: .day, value: -(Constants.totalDays - 1), to: endDate)!

        var dates: [Date?] = []

        // Build grid: 7 rows (days) x 13 columns (weeks)
        for dayOfWeek in 0..<Constants.daysPerWeek {
            for week in 0..<Constants.weeksToShow {
                let dayOffset = week * 7 + dayOfWeek
                let date = calendar.date(byAdding: .day, value: dayOffset, to: startDate)!

                if date > today {
                    dates.append(nil)
                } else {
                    dates.append(date)
                }
            }
        }

        return dates
    }

    static func formatDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: date)
    }

    // MARK: - Logout

    func logout() {
        tokenStorage.clear()
        activityCache.clear()
        WidgetCenter.shared.reloadTimelines(ofKind: "StravaActivityWidget")
    }
}

// MARK: - Errors

enum RepositoryError: Error, LocalizedError {
    case notLoggedIn
    case tokenRefreshFailed
    case noAccessToken

    var errorDescription: String? {
        switch self {
        case .notLoggedIn:
            return "Not logged in to Strava"
        case .tokenRefreshFailed:
            return "Failed to refresh authentication token"
        case .noAccessToken:
            return "No access token available"
        }
    }
}
